### PR TITLE
INT-6857 automate upgrade of charts

### PR DESCRIPTION
**Pull request by jflinchbaugh**
#### Description of Change
Automate upgrade of charts to make it all happen from jenkins. It requires as parameters: the chart, chart version, and application version that's being used for the upgrade.

JIRA: https://issues.sonatype.org/browse/INT-6857
Build: https://jenkins.ci.sonatype.dev/job/integrations/job/cloud/job/Helm3%20Charts/job/Test%20Release%20Build/
(It failed on Post Deploy because the commit for upgrade is not on main yet.)

Integrations is a little short-handed this week, so I've added a few past teammates for a quick review. Thanks.

### CODE REVIEW

File: Jenkinsfile-Release [hunk 1]
```diff
@@ -12,22 +12,57 @@
...
+])
+
 dockerizedBuildPipeline(
   prepare: {
+    if (! params.chart) {
+      error('Chart parameter is required.')
+    }
+    if (! params.chartVersion || ! params.appVersion) {
+      error('Version parameters are required.')
+    }
```
Comment (jflinchbaugh) [inline Jenkinsfile-Release:55]: bail out if this is being run with parameters missing.

File: Jenkinsfile-Release [hunk 2]
```diff
@@ -12,22 +12,57 @@
...
 
+final chartLocation = [
+  'nexus-iq-server': 'nexus-iq',
+  'nexus-repository-manager': 'nexus-repository-manager',
+]
+
+properties([
+  parameters([
+    choice(
+      choices: ['', 'nexus-iq-server', 'nexus-repository-manager'],
```
Comment (jflinchbaugh) [inline Jenkinsfile-Release:33]: default the choice to '', so the user very intentionally needs to choose a chart, and won't accidentally try to upgrade nexus-iq-server to a nexus-repository-manager version.

File: Jenkinsfile-Release [hunk 3]
```diff
@@ -12,22 +12,57 @@
...
+    }
     githubStatusUpdate('pending')
   },
   buildAndTest: {
-    sh './build.sh'
+    final branch = gitBranch(env)
+    runSafely 'git config user.email "sonatype-ci@sonatype.com"'
+    runSafely 'git config user.name "Sonatype CI"'
+    runSafely "git checkout ${branch}"
+    runSafely "./upgrade.sh charts/${chartLocation[params.chart]} ${params.chartVersion} ${params.appVersion}"
```
Comment (jflinchbaugh) [inline Jenkinsfile-Release:63]: this is new: before running the build, use those versions we're given to run the previously-existing upgrade script.

File: Jenkinsfile-Release [hunk 4]
```diff
@@ -36,24 +71,20 @@ dockerizedBuildPipeline(
   deploy: {
     final branch = gitBranch(env)
 
-    runSafely 'git config user.email "sonatype-ci@sonatype.com"'
-    runSafely 'git config user.name "Sonatype CI"'
-    runSafely "git checkout ${branch}"
     runSafely 'git add docs'
-    runSafely 'git commit -m "Release Update"'    
+    runSafely "git commit -m 'Release Update for ${params.chart} ${params.chartVersion}'"
```
Comment (jflinchbaugh) [inline Jenkinsfile-Release:75]: i thought it would be nice to tell us what we're upgrading in our message.

File: Jenkinsfile-Release [hunk 5]
```diff
@@ -12,22 +12,59 @@
...
+    }
+    if (! params.appVersion) {
+      error('The appVersion is required.')
+    }
     githubStatusUpdate('pending')
   },
   buildAndTest: {
-    sh './build.sh'
+    final branch = gitBranch(env)
+    runSafely 'git config user.email "sonatype-ci@sonatype.com"'
```
Comment (kellyrob99) [inline Jenkinsfile-Release:62]: you can use this convenience mechanism now: https://docs.sonatype.com/display/CDI/sonatypeZionGitConfig

File: Jenkinsfile-Release [hunk 6]
```diff
@@ -88,9 +121,29 @@ String verifyIndexYamlAndTarFile(String chart) {
...
+String calculateChartVersion(final String chartVersion, final String appVersion) {
+  if (chartVersion) {
+    return chartVersion
+  }
+
+  if (! appVersion) {
+    error 'Failed to calculate chartVersion with no appVersion.'
+  }
+
+  final versionParts = appVersion.split(/\./)
```
Comment (kellyrob99) [inline Jenkinsfile-Release:140]: should be able to utilize https://docs.sonatype.com/display/CDI/parseVersionString

**Comment by jflinchbaugh** [comment]
thanks @kellyrob99 for the review. your suggestions all worked fine. 👍

**Pull request merged**